quinn-proto/src/connection.rs

```diff
@@ -2032,46 +2051,33 @@ impl Connection {
             // Retry packets are not encrypted and have no packet number
             return Ok(None);
         }
-        let (key_phase, number) = match packet.header {
-            Header::Short {
-                key_phase, number, ..
-            }
-                if !handshake =>
-            {
-                (key_phase, number)
-            }
+        let number = match packet.header {
+            Header::Short {number, ..} if !handshake => number,
             Header::Initial { number, .. } | Header::Long { number, .. } if handshake => {
-                (false, number)
+                number
             }
             _ => {
                 return Err(None);
             }
         };
         let number = number.expand(self.rx_packet + 1);
-        if key_phase != self.key_phase {
-            if number <= self.rx_packet {
-                // Illegal key update
-                return Err(Some(TransportError::PROTOCOL_VIOLATION));
-            }
-            let new = self.crypto.as_mut().unwrap().update(self.side);
-            new.decrypt(number, &packet.header_data, &mut packet.payload)
-                .map_err(|()| None)?;
-
-            let old = mem::replace(self.crypto.as_mut().unwrap(), new);
-            self.prev_crypto = Some((number, old));
-            self.key_phase = !self.key_phase;
-            Ok(Some(number))
-        } else {
-            let crypto = match (handshake, &self.prev_crypto) {
-                (true, _) => &self.handshake_crypto,
-                (false, &Some((boundary, ref prev))) if number < boundary => prev,
-                _ => self.crypto.as_ref().unwrap(),
-            };
-            crypto
-                .decrypt(number, &packet.header_data, &mut packet.payload)
-                .map_err(|()| None)?;
-            Ok(Some(number))
-        }
+        let crypto = match (handshake, &self.prev_crypto) {
+            (true, _) => &self.handshake_crypto,
+            (false, &Some((boundary, ref prev))) if number < boundary => prev,
+            _ => &self.crypto.as_ref().unwrap(),
+        };
+
+        crypto
+            .decrypt(number, &packet.header_data, &mut packet.payload)
+            .map_err(|()| None)?;
+        Ok(Some(number))
+    }
+
+    pub(crate) fn update_keys(&mut self) {
```

Comment:
This seems to be only used in the tests? If so, it probably should live in the `tests` module.